### ISSUE TITLE
Analysis doesn't mark function after conditional loop

### ISSUE THREAD

**Issue by justanotheranonymoususer** [x64dbg/x64dbg#2536]
See the image below. It looks like the code simply treats this jump, or all jumps, as unconditional? Otherwise it would mark the function further. I've had similar mis-marking problems in the past, but this time it's a simple case so I report it, perhaps fixing it will fix the complex ones as well.

![image](https://user-images.githubusercontent.com/5781692/101266289-1b51e880-3756-11eb-8bcc-dd3b277a83c7.png)

**Comment by mrexodia** [comment]
For the ground truth, check the graph view. What does it show there?

**Comment by justanotheranonymoususer** [comment]
Yes, graph view rarely fails :) but the mis-marking also bothers in that if I click "g" in the non-marked spot, it shows from the middle of the func,
P.S. I can upload the file if needed and provide the address.

![image](https://user-images.githubusercontent.com/5781692/101266403-2a856600-3757-11eb-8a05-fc934835372c.png)

**Comment by mrexodia** [comment]
Yeah, please do. Likely there is a conflict with the function ranges (eg
first it created the small range and then the bigger range fails because it
cannot overlap).

With regards to the G problem in the middle of the function, on the latest
versions in many x64 exes it will now work (because I used the
runtime_function info to try to get to the start of the function before
graphing)

**Comment by justanotheranonymoususer** [comment]
Here's this file: [will expire in 30 days](https://easyupload.io/8a4xhe), [hash and info](https://www.virustotal.com/gui/file/46351bd350799dc196481cbe2b26f628b489a280a9e2f49bace71930f3dc80d3/detection)

Function from MS PDB:
`private: bool __cdecl LOCATOR::FLocateCvFilePathHelper(unsigned short const * __ptr64,unsigned short const * __ptr64,unsigned __int64,bool) __ptr64`

RVA: FA530

Thanks.

**Comment by mrexodia** [comment]
![image](https://user-images.githubusercontent.com/2458265/101266786-adf08880-3752-11eb-9ea2-ce37ddab6cb8.png)

No issues for me.

**Comment by justanotheranonymoususer** [comment]
Hmm, I downloaded the latest build, even though mine is quite recent, and you're right, it looks fine. Except I don't see the loops, is that a plugin, or some option?

Still, the problem can be seen in other places, e.g. the `_raise_exc_ex` function: https://i.imgur.com/KIkAjCt.png

Or, here's a simpler case: this function:
`private: unsigned __int64 __cdecl objf::ImageImpl::AddrByRva(unsigned long)const __ptr64`

![image](https://user-images.githubusercontent.com/5781692/101266884-e6489480-375b-11eb-99d0-542f8cd40480.png)

**Comment by mrexodia** [comment]
Probably you’re trying to analyze all functions at once which isn’t really
supported anymore. I just analyze single functions I’m interested in with A

**Comment by justanotheranonymoususer** [comment]
Well, yeah, I'm using Ctrl+A since forever, it's sad news that it's not supported anymore :( any plans on reviving it? It's much more convenient to have all functions analyzed at once.

**Comment by justanotheranonymoususer** [comment]
Wait, does that mean that x-ref detection in whole module is not supported too? That's kinda essential and sad.

**Comment by mrexodia** [comment]
It’s supported, just wasn’t ever any good :)

The main issue is that the analysis is very slow if you do it on all
functions and there isn’t much benefit to analyzing everything, because you
cannot look at everything at once anyway. What is your use case for having
all functions analyzed?

The algorithm used by ‘A’ is not too bad, but if you try to run it on a few
functions it will take way too long. If I ever manage to get it fast it’s 1
line of code to analyze recursively.

**Comment by mrexodia** [comment]
Just run the command “analx” if you need those, it will be instant (and of
course supported). I could try to fix the ctrl+a but the analysis algorithm
is fundamentally flawed (since it’s based on ranges and not on the control
flow graph)

**Comment by justanotheranonymoususer** [comment]
> What is your use case for having all functions analyzed?

It's just convenient to be able to press Ctrl+A and be done with the analysis. Then I can debug a binary, use x-ref, go to graph view, see labels such as "func+0x10" and not "module+0x123410" etc. etc., all benefits of analysis.

It's true that I don't look at all of the module, but clicking "A" all the time is not very convenient too.

How about, when pressing Ctrl+A, doing "analx" and then doing the rest of the analysis completely in the background, such that even if it's slow, the debugger can be meanwhile used? Currently Ctrl+A blocks some input.

By the way, Another issue of "analx" + "A" per function, that I see now, aside from being clumsy, is that using "A" removes a function x-refs.
https://i.imgur.com/yTjsw4s.gif

Also, there's no way to remove all analysis info from a module, right?

**Comment by mrfearless** [comment]
I recommend using https://github.com/ThunderCls/xAnalyzer as it can be set to auto analyse the debugee on first load

**Comment by justanotheranonymoususer** [comment]
I like xAnalyzer, but it's slower than Ctrl+A, has to be installed so more hassle on random PCs, and also blocks debugging while the analysis is in progress.

**Comment by coldwires** [comment]
Another case of this with exact numbers, in case a fresh one helps.

MSVC x64 binary with a table of `jmp` thunks below the first `.pdata` entry. After `Ctrl+A` on the module, one function record covers 92% of `.text`:

| | RVA |
| --- | --- |
| `func.start` | `28EC` |
| `func.end` | `7B2D9` |
| record size | 494,061 bytes |
| `.text` size | 536,795 bytes |

`28EC` is one of those thunks, and the first `.pdata` entry is at `7290`:

```
28EC:  E9 BB 89 07 00      jmp 7B2AC
       28EC + 5 + 789BB = 7B2AC
```

`.pdata` entry 1481 describes the real function as `7B2AC..7B2E0`, 52 bytes. So the correct answer is 52 bytes and the record is 494 KB, because the interval is anchored on the thunk rather than on the function it jumps to.

This lines up with your comment earlier:

> I could try to fix the ctrl+a but the analysis algorithm is fundamentally flawed
> (since it's based on ranges and not on the control flow graph)

The two blocks are as far apart as they can be inside the section, so the interval swallows everything between them, including roughly 1500 unrelated functions and all the `int3` padding.

Everything reading the function database inherits it. `DbgGetFunctionTypeAt` returns `FUNC_MIDDLE` across the whole span, so the brackets in the disassembly are drawn over all of it, and `func.start` / `func.end` report the same range to scripts.

`exanalyse` on the same binary gives the correct 52 byte function, so the data needed to get this right is present in the file.

One narrower thing that would fix this particular shape without touching the range representation: when a candidate entry is a single unconditional `jmp` to a target in the same module, follow it and anchor the function on the target instead. That leaves the wider
problem in place but removes the case where one record covers most of a section.

Happy to share the binary if a sample is useful.

**Comment by mrexodia** [comment]
The `exanalyze` command is exactly meant for this, pdata ranges are not an accurate way of analyzing functions in the general case (since you can put whatever you like there) so the regular analysis command doesn't use them. The stubs thing is a bug, but like I said earlier I don't think it makes much sense to support anything but `exanalyze` and the `A` command to analyze a single function. If it's an easy patch to detect the thunks it might be fine though, but actual function boundary analysis is just an expensive operation.
